src/bam/md_align.rs

```diff
@@ -0,0 +1,1465 @@
+         "GCTGGTGTTAAGGCTGAATGATTCGATCAAAAGCGAAGTATTTTTTT-TTCCTAGTTTTTATGTAAAAATATATTCATTTTTTTCTCTACGCAGTTGCTGCAGTGTTATAAGAAAAATAGAATGAAATAAGATAAAGCTAGACGAATTGTG",
+         "CGTCAGCATCGTGGTAGTCAAAAAGTTCATCTGCACCGTACTCTTTCAACAATTTTTCATGTTTACGAGAAGCAACGACGATGATCTTGCTGAAACCGTTTAG-TTTTTTTGCCAATTGAATAAGCATCTGGCCAACAGCAGTGGCACCAC",
+         "ACAAACCATGAATCCTTTGTAGGATCAATAAACGATCCTGAAGATGGACCTGCTGTTGCACATACTGTTTATTTGGCTGCCTTGGTATACCTCGTGTTTTTCGTATTCTGTGGGTTCCAAGTTTACCTAGCCAGAAGAAAACCTTCGATCG",
+         "CGACATATGGAGATACTTTATTTCCTTTTCTTAATTATTAACGTATACCTATAAATTAACAAAGTATCTAAACAAAATACATAAGTGTACTCAAACTGAGTAGAATCGTCGATTAAACTTCCTTCTCCTTTTAAAAATTAAAAACAGCAAA"];
+
+    #[test]
+    fn test_md_align() {
+        let mut bam = Reader::from_path(&Path::new("test/test_md_align.bam"))
+            .ok()
+            .expect("Error opening file.");
+
+        for (i, record) in bam.records().enumerate() {
+            let mut rec = record.ok().expect("Error reading BAM record");
+            assert_eq!(rec.qname(), TEST_READ_NAMES[i]);
+
+            assert_eq!(
+                MatchDesc::get_md(&rec).ok().expect("No MD field"),
+                TEST_MD[i]
+            );
+            assert_eq!(rec.cigar().to_string(), TEST_CIGAR[i]);
+
+            let ap_iter = AlignPosIter::new_from_record(&rec)
+                .ok()
+                .expect("Creating AlignPosIter");
+            let ap_res: Result<Vec<AlignPos>, MDAlignError> = ap_iter.collect();
+            let aps = ap_res.ok().expect("Unable to create Vec<AlignPos>");
+            let ap_read_line: String = aps.iter().map(|ap| ap.read_line_char()).collect();
+            assert_eq!(ap_read_line, TEST_READ_LINE[i]);
+            let ap_ref_line: String = aps.iter().map(|ap| ap.ref_line_char()).collect();
+            assert_eq!(ap_ref_line, TEST_REF_LINE[i]);
+
+            let rap_iter = RecAlignPosIter::new(&rec)
+                .ok()
+                .expect("Creating RecAlignPosIter");
+            let rap_res: Result<Vec<RecAlignPos>, MDAlignError> = rap_iter.collect();
+            let raps = rap_res.ok().expect("Unable to create Vec<RecAlignPos>");
+            let rap_read_line: String = raps.iter().map(|rap| rap.read_line_char()).collect();
+            assert_eq!(rap_read_line, TEST_READ_LINE[i]);
+            let rap_ref_line: String = raps.iter().map(|rap| rap.ref_line_char()).collect();
+            assert_eq!(rap_ref_line, TEST_REF_LINE[i]);
+
+            let md_iter = MDPosIter::new_from_record(&rec)
+                .ok()
+                .expect("Creating MDPosIter");
+            let alignment = md_iter
+                .collect_into_alignment()
+                .ok()
+                .expect("collecting into alignment");
+            let a_cigar = alignment.cigar(false);
+            assert_eq!(a_cigar, TEST_ALIGN_CIGAR[i]);
```

Comment:
I like the idea of an API like that, and I'll try to play around with it a bit. For my own use of CIGAR+MD parsing, I want to iterate over alignment positions and work with the alignment itself, rather than just reconstruct the reference sequence, so I'd like to keep that functionality available.

The other complication is that reference reconstruction depends on the read sequence in the`rec` (as well as CIGAR + MD) -- the MD string only contains the sequence information that differs between the two.

Reply:
Yes, I forgot about that. So then reference_sequence should rather take the record as an argument. I think your iteration approach could be achieved by providing an iterator over cigar plus reference sequence. This would be more general, in the sense that it can be used both in case of MD and in case of BAM+FASTA.

Reply:
Ok, this is a nice improvement, thanks!
Still, the same kind of iteration could be achieved without MD, given the reference sequence from a FASTA file, right?
So, maybe we can still get a common API for both cases.

Reply:
My point is, an Iterator over `CigarMDPos` can also be implemented with `bam::Record` + `bio::fasta::IndexedReader`. So, I think it should be sufficient for this PR to simply rename `CigarMDPos` to `CigarRefPos`. Then, it is your call whether you want to implement that other iterator in here or whether I or somebody else shall do that later.